src/history/HistoryArchiveReportWork.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "history/HistoryArchiveReportWork.h"
+#include "historywork/GetHistoryArchiveStateWork.h"
+#include "util/Logging.h"
+#include <Tracy.hpp>
+#include <fmt/format.h>
+
+namespace stellar
+{
+HistoryArchiveReportWork::HistoryArchiveReportWork(
+    Application& app,
+    std::vector<std::shared_ptr<HistoryArchive>> const& archives)
+    : Work(app, "history-archive-report-work", BasicWork::RETRY_NEVER)
+    , mGetHistoryArchiveStateWorks()
+    , mAdded(false)
+    , mArchives(archives)
+{
+}
+
+BasicWork::State
+HistoryArchiveReportWork::doWork()
+{
+    ZoneScoped;
+
+    if (!mAdded)
+    {
+        for (auto const& archive : mArchives)
+        {
+            std::shared_ptr<GetHistoryArchiveStateWork> p =
+                addWork<GetHistoryArchiveStateWork>(
+                    0, archive, "archive-report-" + archive->getName(),
+                    BasicWork::RETRY_NEVER);
+            mGetHistoryArchiveStateWorks.push_back(p);
+            mChildren.push_back(p);
+        }
+        mAdded = true;
+    }
+
+    if (allChildrenDone())
+    {
+        // If some work fails quickly and gets cleaned up,
```

Comment:
I think it would be clearer to say something like "Work class provides no guarantees on its stored children lifetime after each child completes, so we can't rely on allChildrenSuccessful, as works might have gotten cleaned up; instead, we store each child work and query its status directly". Something along those lines. What's important to mention here is that we can't rely on the internals of the Work class.

Reply:
Fixed!